cassiopeia/solver/SpectralNeighborJoiningSolver.py

```diff
@@ -0,0 +1,318 @@
+            ] = svd2_val
+
+        np.fill_diagonal(lambda_matrix_arr, np.inf)
+
+        # convert array to dataframe
+        lambda_matrix_df = pd.DataFrame(
+            lambda_matrix_arr, index=node_names, columns=node_names
+        )
+
+        return lambda_matrix_df
+
+    def find_cherry(self, dissimilarity_map: np.ndarray) -> Tuple[int, int]:
+        """Finds a pair of samples to join into a cherry.
+
+        With dissimilarity_map being the lambda matrix, this method finds the
+            argmin pair of subsets of the lambda matrix.
+
+        Args:
+            dissimilarity_matrix: Lambda matrix
+
+        Returns:
+            A tuple of integers representing rows in the
+                dissimilarity matrix to join.
+        """
+
+        return np.unravel_index(
+            np.argmin(dissimilarity_map, axis=None), dissimilarity_map.shape
+        )
+
+    def _compute_svd2(
+        self, pair: Tuple[int, int], lambda_indices: List[List[int]]
+    ) -> float:
+        """Computes the second largest singular value of a pair of
+        subset's RA matrix.
+
+        Args:
+            pair: pair of indices i and j where i > j.
+            lambda_indices: the list of subsets for
+                which 'pair' refers to.
+
+        Returns:
+            The second largest singular value of the pair's RA matrix.
+        """
+        i_idx, j_idx = pair
+        i_sub, j_sub = lambda_indices[i_idx], lambda_indices[j_idx]
+
+        # get combined pair of subsets
+        a_subset = [*i_sub, *j_sub]
+
+        # get complement
+        a_comp = lambda_indices.copy()
+        a_comp.pop(i_idx)
+        a_comp.pop(j_idx)
+        a_comp_flat = list(itertools.chain.from_iterable(a_comp))
+
+        # reconstruct RA matrix
+        RA_matrix = self._similarity_map.values[np.ix_(a_subset, a_comp_flat)]
+
+        # get second largest SVD if available, first if not.
+        s = svd(RA_matrix, compute_uv=False, check_finite=False)
+        svd2_val = s[:2][-1]
+
+        return svd2_val
+
+    def update_dissimilarity_map(
+        self,
+        similarity_map: pd.DataFrame,
+        cherry: Tuple[str, str],
+        new_node: str,
+    ) -> pd.DataFrame:
+        """Updates the lambda matrix using the pair of subsets from find_cherry.
+
+        Args:
+            similarity_map: lambda matrix to update
+            cherry1: One of the children to join.
+            cherry2: One of the children to join.
+            new_node: New node name to add to the dissimilarity map
+
+        Returns:
+            An updated lambda matrix.
+        """
+        # get cherry nodes in index of lambda matrix
+        i, j = (
+            np.where(similarity_map.index == cherry[0])[0][0],
+            np.where(similarity_map.index == cherry[1])[0][0],
+        )
+
+        # modify names
+        node_names = similarity_map.index.values
+        node_names = np.append(node_names, new_node)
+        node_names = np.delete(node_names, [i, j])
```

Comment:
I seem to recall that `np.append` and `np.delete` are actually quite slow, and that's sometimes better to just do simple list concatenations. Can you confirm that it's better to use numpy operations?

Reply:
np.concatenate is better than np.append, so i'll change it to that! But converting the array to a list then back to array takes a lot longer than just direct numpy operations.